File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:

        left, right = 0, len(numbers) - 1

        while left < right:

            while left < right and numbers[left] == numbers[left+1]:
                left += 1
            
            while left < right and numbers[right-1] == numbers[right]:
                right -= 1
            
            curr = numbers[left] + numbers[right]

            if target == curr:
                return [left+1, right+1]
            
            if target > curr:
                left += 1
            else:
                right -= 1
        
        return [-1,-1]

File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int up = 0, down = matrix.size() - 1;
        int left = 0, right = matrix[0].size() - 1;

        while (up < down){
            int mid = up + (down - up)/2 + 1;

            if (matrix[mid][0] > target) down = mid - 1;
            else up = mid;
        }

        while (left <= right){
            int mid = left + (right - left)/2;

            if (matrix[up][mid] == target) return true;
            else if (matrix[up][mid] > target) right = mid - 1;
            else left = mid + 1;
        }

        return false;
        
    }
};
